nlp_utilities.jl

```diff
@@ -0,0 +1,236 @@
+            continue  # Skip variable bounds
+        end
+        for ci in all_constraints(model, F, S)
+            push!(rows, ci)
+            object = constraint_object(ci)
+            MOI.Nonlinear.add_constraint(nlp, object.func, object.set)
+        end
+    end
+    MOI.Nonlinear.set_objective(nlp, objective_function(model))
+    return nlp, rows
+end
+
+"""
+    fill_off_diagonal(H)
+
+Filling the off-diagonal elements of a sparse matrix to make it symmetric.
+"""
+function fill_off_diagonal(H)
+    ret = H + H'
+    row_vals = SparseArrays.rowvals(ret)
+    non_zeros = SparseArrays.nonzeros(ret)
+    for col in 1:size(ret, 2)
+        for i in SparseArrays.nzrange(ret, col)
+            if col == row_vals[i]
+                non_zeros[i] /= 2
+            end
+        end
+    end
+    return ret
+end
+
+"""
+    compute_optimal_hessian(evaluator::MOI.Nonlinear.Evaluator, rows::Vector{ConstraintRef}, x::Vector{VariableRef})
+
+Compute the optimal Hessian of the Lagrangian.
+"""
+function compute_optimal_hessian(evaluator::MOI.Nonlinear.Evaluator, rows::Vector{ConstraintRef}, x::Vector{VariableRef})
+    hessian_sparsity = MOI.hessian_lagrangian_structure(evaluator)
+    I = [i for (i, _) in hessian_sparsity]
+    J = [j for (_, j) in hessian_sparsity]
+    V = zeros(length(hessian_sparsity))
+    MOI.eval_hessian_lagrangian(evaluator, V, value.(x), 1.0, dual.(rows))
+    H = SparseArrays.sparse(I, J, V, length(x), length(x))
+    return Matrix(fill_off_diagonal(H))
+end
+
+"""
+    compute_optimal_jacobian(evaluator::MOI.Nonlinear.Evaluator, rows::Vector{ConstraintRef}, x::Vector{VariableRef})
+
+Compute the optimal Jacobian of the constraints.
+"""
+function compute_optimal_jacobian(evaluator::MOI.Nonlinear.Evaluator, rows::Vector{ConstraintRef}, x::Vector{VariableRef})
+    jacobian_sparsity = MOI.jacobian_structure(evaluator)
+    I = [i for (i, _) in jacobian_sparsity]
+    J = [j for (_, j) in jacobian_sparsity]
+    V = zeros(length(jacobian_sparsity))
+    MOI.eval_constraint_jacobian(evaluator, V, value.(x))
+    A = SparseArrays.sparse(I, J, V, length(rows), length(x))
+    return Matrix(A)
+end
+
+"""
+    compute_optimal_hess_jac(evaluator::MOI.Nonlinear.Evaluator, rows::Vector{ConstraintRef}, x::Vector{VariableRef})
+
+Compute the optimal Hessian of the Lagrangian and Jacobian of the constraints.
+"""
+function compute_optimal_hess_jac(evaluator::MOI.Nonlinear.Evaluator, rows::Vector{ConstraintRef}, x::Vector{VariableRef})
+    hessian = compute_optimal_hessian(evaluator, rows, x)
+    jacobian = compute_optimal_jacobian(evaluator, rows, x)
+    
+    return hessian, jacobian
+end
+
+"""
+    all_primal_vars(model::Model)
+
+Get all the primal variables in the model.
+"""
+all_primal_vars(model::Model) = filter(x -> !is_parameter(x), all_variables(model))
+
+"""
+    all_params(model::Model)
+
+Get all the parameters in the model.
+"""
+all_params(model::Model) = filter(x -> is_parameter(x), all_variables(model))
```

Comment:
We probably want to either consider `EqualTo` constraints in the Jacobian or consider fixed variables as parameters here. I think I would prefer the former for now, as it is more explicit (although I can maybe see the argument/use case for allowing derivatives with respect to fixed variables eventually).

Reply:
> I think I would prefer the former

Me too, that way the user can choose which parameters he wants to diff wrt. But I am open to requests.

Reply:
solved considering the equalto